### PR TITLE
pup: rename attr, go-pup -> pup, clean up, 20160425-e76307d -> 0.4.0, add meta

### PR DESCRIPTION
**Pull request by lukateras**
###### Motivation for this change

pup is a very popular tool ([4084 stars on GitHub](https://github.com/ericchiang/pup)) and in my opinion deserves a three-letter attrname and an update to a stable version :-)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -4136,6 +4134,8 @@ with pkgs;
     libXNVCtrl = linuxPackages.nvidia_x11.settings.libXNVCtrl;
   };
 
+  pup = callPackage ../development/tools/pup { };
```
Comment (orivej) [inline pkgs/top-level/all-packages.nix]: Please move it to the development/tools (`### DEVELOPMENT / TOOLS`) section.

File: pkgs/development/tools/pup/default.nix [hunk 2]
```diff
@@ -1,15 +1,23 @@
...
+    url = "https://${goPackagePath}";
+    sha256 = "0mnhw0yph5fvcnrcmj1kfbyw1a4lcg3k9f6y28kf44ihlq8h1dfz";
+  };
+
+  meta = with lib; {
+    description = "Streaming HTML processor/selector";
+    homepage = "https://${goPackagePath}";
+    license = licenses.mit;
+    maintainers = with maintainers; [ yegortimoshenko ];
+    platforms = platforms.all;
```
Comment (orivej) [inline pkgs/development/tools/pup/default.nix]: FYI `buildGoPackage` automatically sets platforms, but setting it explicitly is also fine. (I checked again to see if it automatically sets homepage, but it does not. You may as well make it do in `pkgs/development/go-modules/generic/default.nix`.)

Comment (yegortimoshenko) [reply pkgs/development/tools/pup/default.nix]: Sure!

**Comment by yegortimoshenko** [comment]
All fixed!

**Pull request merged**